Running testng-failed.xml correctly runs failed test in child class but incorrectly runs all tests in base class

**TestNG version - 6.14.3
IDE - IntelliJ**

# Explanation -

I have tests distributed across child class and a base class that it extends from. Both the classes are in different packages.

**DataProviderTest.class**

```java
package com.rambo.rambokart;

import org.testng.Assert;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import com.rambo.zamboKart.TestBase;

public class DataProviderTest extends TestBase {

	private PrimeNumberChecker primeNumberChecker;

	@BeforeMethod
	public void initialize() {
		primeNumberChecker = new PrimeNumberChecker();
	}

	@DataProvider(name = "test1")
	public static Object[][] primeNumbers() {
		return new Object[][] {{2, true}, {6, false}, {19, true}, {22, true}, {23, false}};
	}

	@Test(dataProvider = "test1")
	public void testPrimeNumberChecker(Integer inputNumber, Boolean expectedResult) {
		System.out.println(inputNumber + " " + expectedResult);
		Assert.assertEquals(expectedResult, primeNumberChecker.validate(inputNumber));
	}
}
```

**TestBase.class**

```java
package com.rambo.zamboKart;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class TestBase {

	@Test(dataProvider = "test2")
	public void testNumberEquality(Integer i1, Integer i2) {
		Assert.assertEquals(i1,i2);
	}

	@DataProvider(name = "test2")
	public static Object[][] numbers() {
		return new Object[][] {{2, 2}, {6, 6}, {19, 18}, {22, 21}, {23, 23}};
	}
}
```

# Suite File - MyTestNG.xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Main Test Suite" verbose="2">

    <test name="DataProvider Tests">
        <classes>
            <class name="com.rambo.rambokart.DataProviderTest" />
        </classes>
    </test>
</suite>
```

Running the above suite file, will run all the tests from DataProviderTest and TestBase class as shown below - 

![image](https://user-images.githubusercontent.com/16848396/46689863-ad312b80-cbb5-11e8-8e61-8997a5e9a726.png)


Now we have following testNG-failed.xml created - 

# testNG-failed.xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="Failed suite [Main Test Suite]">

  <test thread-count="5" verbose="2" name="DataProvider Tests(failed)">
    <classes>
      <class name="com.rambo.rambokart.DataProviderTest">
        <methods>
          <include name="testNumberEquality" invocation-numbers="2 3"/>
          <include name="initialize"/>
          <include name="testPrimeNumberChecker" invocation-numbers="3 4"/>
        </methods>
      </class>
    </classes>
  </test>
</suite>
```

Running this testNG-failed xml gives following output - 

![image](https://user-images.githubusercontent.com/16848396/46689952-ed90a980-cbb5-11e8-8252-bb9a5b17adb1.png)

**Actual** - This test run also runs tests from TestBase class that were successful in their initial run. Note that only failed tests from DataProviderTest class is being executed. SO, behavior is correct for child class but not for class being extended.

**Expected** - Only failed test from classes DataProviderTest and TestBase should run.

@juherr @krmahadevan 

As suggested, I tried running my suite file - testng.xml with TestNG command line. Then I executed TestNG-failed.xml with TestNG command line.

I saw following behavior (which is same as running with IntelliJ IDE)

**TestNG.xml**

```
nishant-macOS:resources nishant$ java -classpath "/Users/nishant/mylib/jcommander-1.72.jar:/Users/nishant/mylib/rambo-kart-0.0.1-SNAPSHOT.jar:/Users/nishant/mylib/testng-7.0.0-beta1.jar:/Users/nishant/mylib/rambo-kart-0.0.1-SNAPSHOT-tests.jar:." org.testng.TestNG testNG.xml 
...
... TestNG 7.0.0-beta1 by Cédric Beust (cedric@beust.com)
...

2 true
6 false
19 true
22 true
23 false
PASSED: testPrimeNumberChecker(2, true)
PASSED: testPrimeNumberChecker(6, false)
PASSED: testPrimeNumberChecker(19, true)
PASSED: testNumberEquality(2, 2)
PASSED: testNumberEquality(6, 6)
PASSED: testNumberEquality(23, 23)
FAILED: testPrimeNumberChecker(22, true)
java.lang.AssertionError: did not expect to find [false] but found [true]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.rambokart.DataProviderTest.testPrimeNumberChecker(DataProviderTest.java:44)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)

FAILED: testPrimeNumberChecker(23, false)
java.lang.AssertionError: did not expect to find [true] but found [false]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.rambokart.DataProviderTest.testPrimeNumberChecker(DataProviderTest.java:44)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)

FAILED: testNumberEquality(19, 18)
java.lang.AssertionError: did not expect to find [18] but found [19]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.zamboKart.TestBase.testNumberEquality(TestBase.java:29)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)

FAILED: testNumberEquality(22, 21)
java.lang.AssertionError: did not expect to find [21] but found [22]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.zamboKart.TestBase.testNumberEquality(TestBase.java:29)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)


===============================================
    DataProvider Tests
    Tests run: 10, Failures: 4, Skips: 0
===============================================


===============================================
Main Test Suite
Total tests run: 10, Passes: 6, Failures: 4, Skips: 0
===============================================
```

**TestNG-failed.xml file generated**

```XML
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite guice-stage="DEVELOPMENT" verbose="2" name="Failed suite [Main Test Suite]">
  <test thread-count="5" verbose="2" name="DataProvider Tests(failed)">
    <classes>
      <class name="com.rambo.rambokart.DataProviderTest">
        <methods>
          <include name="testNumberEquality" invocation-numbers="2 3"/>
          <include name="initialize"/>
          <include name="testPrimeNumberChecker" invocation-numbers="3 4"/>
        </methods>
      </class>
    </classes>
  </test>
</suite> 
```

**TestNG-failed.xml file execution**

```
nishant-macOS:target nishant$ java -classpath "/Users/nishant/mylib/jcommander-1.72.jar:/Users/nishant/mylib/rambo-kart-0.0.1-SNAPSHOT.jar:/Users/nishant/mylib/testng-7.0.0-beta1.jar:/Users/nishant/mylib/rambo-kart-0.0.1-SNAPSHOT-tests.jar:." org.testng.TestNG testng-failed.xml 
...
... TestNG 7.0.0-beta1 by Cédric Beust (cedric@beust.com)
...

22 true
23 false
PASSED: testNumberEquality(2, 2)
PASSED: testNumberEquality(6, 6)
PASSED: testNumberEquality(23, 23)
FAILED: testNumberEquality(19, 18)
java.lang.AssertionError: did not expect to find [18] but found [19]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.zamboKart.TestBase.testNumberEquality(TestBase.java:29)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)

FAILED: testNumberEquality(22, 21)
java.lang.AssertionError: did not expect to find [21] but found [22]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.zamboKart.TestBase.testNumberEquality(TestBase.java:29)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)

FAILED: testPrimeNumberChecker(22, true)
java.lang.AssertionError: did not expect to find [false] but found [true]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.rambokart.DataProviderTest.testPrimeNumberChecker(DataProviderTest.java:44)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)

FAILED: testPrimeNumberChecker(23, false)
java.lang.AssertionError: did not expect to find [true] but found [false]
        at org.testng.Assert.fail(Assert.java:97)
        at org.testng.Assert.failNotEquals(Assert.java:969)
        at org.testng.Assert.assertEqualsImpl(Assert.java:136)
        at org.testng.Assert.assertEquals(Assert.java:118)
        at org.testng.Assert.assertEquals(Assert.java:563)
        at com.rambo.rambokart.DataProviderTest.testPrimeNumberChecker(DataProviderTest.java:44)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:131)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:658)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:792)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1103)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:140)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:122)
        at org.testng.TestRunner.privateRun(TestRunner.java:739)
        at org.testng.TestRunner.run(TestRunner.java:589)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:398)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:392)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:354)
        at org.testng.SuiteRunner.run(SuiteRunner.java:302)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1145)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1067)
        at org.testng.TestNG.runSuites(TestNG.java:997)
        at org.testng.TestNG.run(TestNG.java:965)
        at org.testng.TestNG.privateMain(TestNG.java:1291)
        at org.testng.TestNG.main(TestNG.java:1259)


===============================================
    DataProvider Tests(failed)
    Tests run: 7, Failures: 4, Skips: 0
===============================================


===============================================
Failed suite [Main Test Suite]
Total tests run: 7, Passes: 3, Failures: 4, Skips: 0
===============================================

```

**Expected** - Only 4 failed test from previous run, that are included in testNG-failed.xml should run (and fail again). However, few already "PASSED" test from previous run are also executed.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-1930: Running testng-failed.xml correctly runs failed test in child class but incorrectly runs all tests in base class (Krishnan Mahadevan)
 Fixed: GITHUB-1987: Retrieve the data provider method's reference from the test method (Krishnan Mahadevan)
 Fixed: GITHUB-1976: Add a proper message when types are missing during annotation parsing (Krishnan Mahadevan)
 Fixed: GITHUB-2000: ThreadPoolExecutor ConcurrentModificationException (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/TestNGMethod.java
+++ b/src/main/java/org/testng/internal/TestNGMethod.java
@@ -2,7 +2,6 @@ package org.testng.internal;
 
 import java.util.Arrays;
 import java.util.Objects;
-import java.util.stream.Collectors;
 import org.testng.IDataProviderMethod;
 import org.testng.IRetryAnalyzer;
 import org.testng.ITestClass;
@@ -71,40 +70,36 @@ public class TestNGMethod extends BaseTestMethod {
     setInvocationNumbers(
         xmlTest.getInvocationNumbers(
             m_method.getDeclaringClass().getName() + "." + m_method.getName()));
-    {
-      ITestAnnotation testAnnotation =
-          AnnotationHelper.findTest(getAnnotationFinder(), m_method.getMethod());
+    ITestAnnotation testAnnotation =
+        AnnotationHelper.findTest(getAnnotationFinder(), m_method.getMethod());
 
-      if (testAnnotation == null) {
-        // Try on the class
-        testAnnotation =
-            AnnotationHelper.findTest(getAnnotationFinder(), m_method.getDeclaringClass());
-      }
-
-      if (null != testAnnotation) {
-        setTimeOut(testAnnotation.getTimeOut());
-        m_successPercentage = testAnnotation.getSuccessPercentage();
-        isDataDriven = doesTestAnnotationHaveADataProvider(testAnnotation);
-
-        setInvocationCount(testAnnotation.getInvocationCount());
-        setThreadPoolSize(testAnnotation.getThreadPoolSize());
-        setAlwaysRun(testAnnotation.getAlwaysRun());
-        setDescription(findDescription(testAnnotation, xmlTest));
-        setEnabled(testAnnotation.getEnabled());
-        setRetryAnalyzer(cloneInstance(testAnnotation.getRetryAnalyzer()));
-        setRetryAnalyzerClass(testAnnotation.getRetryAnalyzerClass());
-        setSkipFailedInvocations(testAnnotation.skipFailedInvocations());
-        setInvocationTimeOut(testAnnotation.invocationTimeOut());
-        setIgnoreMissingDependencies(testAnnotation.ignoreMissingDependencies());
-        setPriority(testAnnotation.getPriority());
-        m_attributes = testAnnotation.getAttributes();
-      }
-
-      // Groups
-      {
-        initGroups(ITestAnnotation.class);
-      }
+    if (testAnnotation == null) {
+      // Try on the class
+      testAnnotation =
+          AnnotationHelper.findTest(getAnnotationFinder(), m_method.getDeclaringClass());
     }
+
+    if (null != testAnnotation) {
+      setTimeOut(testAnnotation.getTimeOut());
+      m_successPercentage = testAnnotation.getSuccessPercentage();
+      isDataDriven = doesTestAnnotationHaveADataProvider(testAnnotation);
+
+      setInvocationCount(testAnnotation.getInvocationCount());
+      setThreadPoolSize(testAnnotation.getThreadPoolSize());
+      setAlwaysRun(testAnnotation.getAlwaysRun());
+      setDescription(findDescription(testAnnotation, xmlTest));
+      setEnabled(testAnnotation.getEnabled());
+      setRetryAnalyzer(cloneInstance(testAnnotation.getRetryAnalyzer()));
+      setRetryAnalyzerClass(testAnnotation.getRetryAnalyzerClass());
+      setSkipFailedInvocations(testAnnotation.skipFailedInvocations());
+      setInvocationTimeOut(testAnnotation.invocationTimeOut());
+      setIgnoreMissingDependencies(testAnnotation.ignoreMissingDependencies());
+      setPriority(testAnnotation.getPriority());
+      m_attributes = testAnnotation.getAttributes();
+    }
+
+    // Groups
+    initGroups(ITestAnnotation.class);
   }
 
   private static boolean doesTestAnnotationHaveADataProvider(ITestAnnotation testAnnotation) {

--- a/src/main/java/org/testng/internal/TestNGMethod.java
+++ b/src/main/java/org/testng/internal/TestNGMethod.java
@@ -1,5 +1,8 @@
 package org.testng.internal;
 
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.stream.Collectors;
 import org.testng.IDataProviderMethod;
 import org.testng.IRetryAnalyzer;
 import org.testng.ITestClass;
@@ -111,22 +114,25 @@ public class TestNGMethod extends BaseTestMethod {
 
   private String findDescription(ITestAnnotation testAnnotation, XmlTest xmlTest) {
     String result = testAnnotation.getDescription();
-    if (result == null) {
-      List<XmlClass> classes = xmlTest.getXmlClasses();
-      for (XmlClass c : classes) {
-        if (c.getName().equals(m_method.getMethod().getDeclaringClass().getName())) {
-          for (XmlInclude include : c.getIncludedMethods()) {
-            if (include.getName().equals(m_method.getName())) {
-              result = include.getDescription();
-              if (result != null) {
-                break;
-              }
-            }
-          }
-        }
-      }
+    if (result != null) {
+      return result;
     }
-    return result;
+    List<XmlClass> classes = xmlTest.getXmlClasses();
+    return classes.stream()
+        .filter(this::classNameMatcher)
+        .flatMap(xmlClass -> xmlClass.getIncludedMethods().stream())
+        .filter(this::methodNameMatcher)
+        .map(XmlInclude::getDescription)
+        .filter(Objects::nonNull)
+        .findFirst().orElse("");
+  }
+
+  private boolean classNameMatcher(XmlClass xmlClass) {
+    return xmlClass.getName().equals(m_method.getMethod().getDeclaringClass().getName());
+  }
+
+  private boolean methodNameMatcher(XmlInclude xmlInclude) {
+    return xmlInclude.getName().equals(m_method.getName());
   }
 
   /** {@inheritDoc} */
@@ -192,12 +198,9 @@ public class TestNGMethod extends BaseTestMethod {
   }
 
   private ITestNGMethod[] clone(ITestNGMethod[] sources) {
-    ITestNGMethod[] clones = new ITestNGMethod[sources.length];
-    for (int i = 0; i < sources.length; i++) {
-      clones[i] = sources[i].clone();
-    }
-
-    return clones;
+    return Arrays.stream(sources)
+        .map(ITestNGMethod::clone)
+        .toArray(ITestNGMethod[]::new);
   }
 
   private static IRetryAnalyzer cloneInstance(IRetryAnalyzer instance) {

--- a/src/main/java/org/testng/internal/TestNGMethod.java
+++ b/src/main/java/org/testng/internal/TestNGMethod.java
@@ -67,9 +67,13 @@ public class TestNGMethod extends BaseTestMethod {
 
   private void init(XmlTest xmlTest) {
     setXmlTest(xmlTest);
-    setInvocationNumbers(
-        xmlTest.getInvocationNumbers(
-            m_method.getDeclaringClass().getName() + "." + m_method.getName()));
+    String className = m_method.getDeclaringClass().getName();
+    Object obj = getInstance();
+    if (obj != null) {
+      className = obj.getClass().getName();
+    }
+    setInvocationNumbers(xmlTest.getInvocationNumbers(className + "." + m_method.getName()));
+
     ITestAnnotation testAnnotation =
         AnnotationHelper.findTest(getAnnotationFinder(), m_method.getMethod());
 
@@ -192,7 +196,7 @@ public class TestNGMethod extends BaseTestMethod {
     return clone;
   }
 
-  private ITestNGMethod[] clone(ITestNGMethod[] sources) {
+  private static ITestNGMethod[] clone(ITestNGMethod[] sources) {
     return Arrays.stream(sources)
         .map(ITestNGMethod::clone)
         .toArray(ITestNGMethod[]::new);

--- a/src/test/java/test/failures/FailuresTest.java
+++ b/src/test/java/test/failures/FailuresTest.java
@@ -1,7 +1,14 @@
 package test.failures;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 import org.testng.TestNG;
 import org.testng.annotations.Test;
+import org.testng.collections.Maps;
 import org.testng.reporters.FailedReporter;
 import org.testng.xml.XmlSuite;
 
@@ -12,6 +19,8 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.HashMap;
 import java.util.Map;
+import test.failures.issue1930.SimpleCliStatus;
+import test.failures.issue1930.TestClassSample;
 
 public class FailuresTest extends BaseFailuresTest {
 
@@ -64,5 +73,35 @@ public class FailuresTest extends BaseFailuresTest {
     tng.run();
 
     verify(tempDirectory, suiteName, expectedParameter);
+  }
+
+  @Test(description = "GITHUB-1930")
+  public void testToEnsureThatWeRunOnlyFailedIterationsFromBaseClass() {
+    File outputDir = createDirInTempDir("testng-tmp-" + System.currentTimeMillis() % 1000);
+    TestNG testng = create(TestClassSample.class);
+    testng.setOutputDirectory(outputDir.getAbsolutePath());
+    testng.setUseDefaultListeners(true);
+    testng.run();
+    String file = outputDir.getAbsolutePath() + File.separator + FailedReporter.TESTNG_FAILED_XML;
+
+    //First iteration of running failed tests.
+    runIteration(outputDir, file);
+    //Second iteration of running failed tests.
+    runIteration(outputDir, file);
+  }
+
+  private static void runIteration(File outputDir, String file) {
+    SimpleCliStatus listener = new SimpleCliStatus();
+    TestNG testng = create();
+    testng.setTestSuites(Collections.singletonList(file));
+    testng.setOutputDirectory(outputDir.getAbsolutePath());
+    testng.setUseDefaultListeners(true);
+    testng.addListener(listener);
+    testng.run();
+
+    Map<String, List<Integer>> expected = Maps.newHashMap();
+    expected.put("testPrimeNumberChecker", Arrays.asList(3, 4));
+    expected.put("testNumberEquality", Arrays.asList(2, 3));
+    assertThat(listener.getFailedTests()).containsAllEntriesOf(expected);
   }
 }

--- a/src/test/java/test/failures/issue1930/PrimeNumberChecker.java
+++ b/src/test/java/test/failures/issue1930/PrimeNumberChecker.java
@@ -1,0 +1,20 @@
+package test.failures.issue1930;
+
+public class PrimeNumberChecker {
+
+  public boolean validate(int number) {
+    if (number == 2 || number == 3) {
+      return true;
+    }
+    if (number % 2 == 0) {
+      return false;
+    }
+    int sqrt = (int) Math.sqrt(number) + 1;
+    for (int i = 3; i < sqrt; i += 2) {
+      if (number % i == 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+}

--- a/src/test/java/test/failures/issue1930/SimpleCliStatus.java
+++ b/src/test/java/test/failures/issue1930/SimpleCliStatus.java
@@ -1,0 +1,26 @@
+package test.failures.issue1930;
+
+import java.util.List;
+import java.util.Map;
+import org.testng.ITestContext;
+import org.testng.ITestListener;
+import org.testng.collections.Maps;
+
+public class SimpleCliStatus implements ITestListener {
+  private Map<String, List<Integer>> failedTests = Maps.newHashMap();
+
+  public Map<String, List<Integer>> getFailedTests() {
+    return failedTests;
+  }
+
+  @Override
+  public void onFinish(ITestContext context) {
+    context.getFailedTests()
+        .getAllResults()
+        .forEach(
+            result -> failedTests
+                .put(result.getMethod().getMethodName(),
+                    result.getMethod().getInvocationNumbers())
+        );
+  }
+}

--- a/src/test/java/test/failures/issue1930/TestBase.java
+++ b/src/test/java/test/failures/issue1930/TestBase.java
@@ -1,0 +1,18 @@
+package test.failures.issue1930;
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class TestBase {
+
+  @Test(dataProvider = "test2")
+  public void testNumberEquality(Integer i1, Integer i2) {
+    Assert.assertEquals(i1, i2);
+  }
+
+  @DataProvider(name = "test2")
+  public static Object[][] numbers() {
+    return new Object[][] {{2, 2}, {6, 6}, {19, 18}, {22, 21}, {23, 23}};
+  }
+}

--- a/src/test/java/test/failures/issue1930/TestClassSample.java
+++ b/src/test/java/test/failures/issue1930/TestClassSample.java
@@ -1,0 +1,27 @@
+package test.failures.issue1930;
+
+import org.testng.Assert;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+public class TestClassSample extends TestBase {
+
+  private PrimeNumberChecker primeNumberChecker;
+
+  @BeforeMethod
+  public void initialize() {
+    primeNumberChecker = new PrimeNumberChecker();
+  }
+
+  @DataProvider(name = "test1")
+  public static Object[][] primeNumbers() {
+    return new Object[][] {{2, true}, {6, false}, {19, true}, {22, true}, {23, false}};
+  }
+
+  @Test(dataProvider = "test1")
+  public void testPrimeNumberChecker(int inputNumber, boolean expectedResult) {
+    Assert.assertEquals(expectedResult, primeNumberChecker.validate(inputNumber));
+  }
+}

